flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapUnionFunction.java

```diff
@@ -0,0 +1,153 @@
+        KeyValueDataType outputType =
+                ((KeyValueDataType) context.getCallContext().getOutputDataType().get());
+        final DataType keyDataType = outputType.getKeyDataType();
+        final DataType valueDataType = outputType.getValueDataType();
+        keyElementGetter =
+                ArrayData.createElementGetter(outputType.getKeyDataType().getLogicalType());
+        valueElementGetter =
+                ArrayData.createElementGetter(outputType.getValueDataType().getLogicalType());
+        keyEqualityEvaluator =
+                context.createEvaluator(
+                        $("element1").isEqual($("element2")),
+                        DataTypes.BOOLEAN(),
+                        DataTypes.FIELD("element1", keyDataType.notNull().toInternal()),
+                        DataTypes.FIELD("element2", keyDataType.notNull().toInternal()));
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        keyEqualityHandle = keyEqualityEvaluator.open(context);
+    }
+
+    public @Nullable MapData eval(@Nullable MapData map1, @Nullable MapData map2) {
+        try {
+            if (map1 == null || map2 == null) {
+                return null;
+            }
+            if (map1.size() == 0) {
+                if (map2.size() == 0) {
+                    return map1;
+                }
+                return map2;
+            }
+            if (map2.size() == 0) {
+                return map1;
+            }
+            return new MapDataForMapUnion(map1, map2);
+        } catch (Throwable t) {
+            throw new FlinkRuntimeException(t);
+        }
+    }
+
+    private class MapDataForMapUnion implements MapData {
+        private final GenericArrayData keysArray;
+        private final GenericArrayData valuesArray;
+
+        public MapDataForMapUnion(MapData map1, MapData map2) throws Throwable {
+            List<Object> keysList = new ArrayList<>();
+            List<Object> valuesList = new ArrayList<>();
+
+            for (int i = 0; i < map1.size(); i++) {
+                keysList.add(keyElementGetter.getElementOrNull(map1.keyArray(), i));
+                valuesList.add(valueElementGetter.getElementOrNull(map1.valueArray(), i));
+            }
+
+            for (int i = 0; i < map2.size(); i++) {
+                final Object key2 = keyElementGetter.getElementOrNull(map2.keyArray(), i);
+                final Object value2 = valueElementGetter.getElementOrNull(map2.valueArray(), i);
+
+                boolean keyExists = false;
+                if (key2 != null) {
+                    for (int j = 0; j < keysList.size(); j++) {
+                        final Object key1 = keysList.get(j);
+                        if (key1 != null && (boolean) keyEqualityHandle.invoke(key1, key2)) {
+                            // If key exists, update the value with the value from the second map
+                            valuesList.set(j, value2);
```

Comment:
nit: if you start `MapDataForMapUnion` with `map2` you don't neet update opeartion

Reply:
good suggestion, thank you, I will change it.